src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs

```diff
@@ -2,44 +2,103 @@
 using System;
 using System.IO;
 using System.Management.Automation;
+using System.Management.Automation.Internal;
 
 namespace Microsoft.PowerShell.Commands
 {
     /// <summary>
-    /// The implementation of the "New-TemporaryFile" cmdlet
+    /// The implementation of the "New-TemporaryFile" cmdlet that has an optional 'Extension' Parameter property.
+    /// If this cmdlet errors then it throws a non-terminating error.
     /// </summary>
     [Cmdlet(VerbsCommon.New, "TemporaryFile", SupportsShouldProcess = true, HelpUri = "https://go.microsoft.com/fwlink/?LinkId=526726")]
     [OutputType(typeof(System.IO.FileInfo))]
     public class NewTemporaryFileCommand : Cmdlet
     {
+        private const string defaultExtension = ".tmp";
+
+        /// <summary>
+        /// Specify a different file extension other than the default one, which is '.tmp'. The period in this parameter is optional.
+        /// </summary>
+        [Parameter(Position = 0)]
+        [ValidateNotNullOrEmpty]
+        public string Extension { get; set; } = defaultExtension;
+
         /// <summary>
-        /// returns a TemporaryFile
+        /// Returns a TemporaryFile.
         /// </summary>
         protected override void EndProcessing()
         {
-            string filePath = null;
-            string tempPath = System.Environment.GetEnvironmentVariable("TEMP");
+            // Check for invalid characters in extension
+            if (Extension.IndexOfAny(Path.GetInvalidFileNameChars()) != -1)
+            {
+                WriteError(
+                    new ErrorRecord(
+                        new ArgumentException(
+                                StringUtil.Format(UtilityCommonStrings.InvalidCharacterInParameter,
+                                    nameof(Extension), Extension)),
+                        "NewTemporaryInvalidArgument",
+                        ErrorCategory.InvalidArgument,
+                        Extension));
+                return;
+            }
+
+            string temporaryFilePath = null;
+            string tempPath = Path.GetTempPath();
+
             if (ShouldProcess(tempPath))
             {
-                try
+                int attempts = 0;
+                bool creationOfFileSuccessful = false;
+
+                // In case the random temporary file already exists, retry 
+                while (attempts++ < 10 && !creationOfFileSuccessful)
                 {
-                    filePath = Path.GetTempFileName();
+                    try
+                    {
+                        temporaryFilePath = Path.GetTempFileName(); // this already creates the temporary file
+                        if (Extension != defaultExtension)
+                        {
```

Comment:
We avoid string comparisons using an operator, instead preferring `string.Equals` and explicitly specify the comparison with either Ordinal or OrdinalIgnoreCase. In this case, Ordinal is probably fine, but we would want a test to make sure File.Move works fine when only the case of the extension differs (e.g. `-Extension .TMP`.)

One other thought - if we compare the extension of `temporaryFilePath` with `Extension`, we can avoid a file move if `GetTempFileName` changed it's algorithm and the new algorithm matches what the user specifies for `Extension` - an unlikely scenario I suppose, but I don't see a downside to this code change.